src/wallet/wallet.cpp

```diff
@@ -6126,54 +6126,134 @@ bool PaymentAddressBelongsToWallet::operator()(const libzcash::SaplingPaymentAdd
+        return PaymentAddressSource::ImportedWatchOnly;
+    } else {
+        return PaymentAddressSource::AddressNotFound;
+    }
 }
 PaymentAddressSource GetSourceForPaymentAddress::operator()(const libzcash::SaplingPaymentAddress &zaddr) const
 {
-    libzcash::SaplingIncomingViewingKey ivk;
-
-    // If we have a SaplingExtendedSpendingKey in the wallet, then we will
-    // also have the corresponding SaplingExtendedFullViewingKey.
-    if (m_wallet->GetSaplingIncomingViewingKey(zaddr, ivk)) {
-        if (m_wallet->HaveSaplingFullViewingKey(ivk)) {
-            // If we have the HD keypath, it's related to the legacy seed
-            if (m_wallet->mapSaplingZKeyMetadata.count(ivk) > 0 &&
-                    m_wallet->mapSaplingZKeyMetadata[ivk].hdKeypath != "") {
-                return LegacyHDSeed;
-            } else if (m_wallet->HaveSaplingSpendingKeyForAddress(zaddr)) {
-                return Imported;
+    auto ufvkSource = this->GetUnifiedSource(zaddr);
+    if (ufvkSource == PaymentAddressSource::AddressNotFound) {
+        libzcash::SaplingIncomingViewingKey ivk;
+
+        // If we have a SaplingExtendedSpendingKey in the wallet, then we will
+        // also have the corresponding SaplingExtendedFullViewingKey.
+        if (m_wallet->GetSaplingIncomingViewingKey(zaddr, ivk)) {
+            if (m_wallet->HaveSaplingFullViewingKey(ivk)) {
+                // If we have the HD keypath, it's related to the legacy seed
+                if (m_wallet->mapSaplingZKeyMetadata.count(ivk) > 0 &&
+                        m_wallet->mapSaplingZKeyMetadata[ivk].hdKeypath != "") {
+                    return PaymentAddressSource::LegacyHDSeed;
+                } else if (m_wallet->HaveSaplingSpendingKeyForAddress(zaddr)) {
+                    return PaymentAddressSource::Imported;
+                } else {
+                    return PaymentAddressSource::ImportedWatchOnly;
+                }
             } else {
-                return ImportedWatchOnly;
+                return PaymentAddressSource::ImportedWatchOnly;
             }
-        } else {
-            return ImportedWatchOnly;
         }
-    } else {
-        return AddressNotFound;
     }
+
+    return ufvkSource;
 }
 PaymentAddressSource GetSourceForPaymentAddress::operator()(const libzcash::UnifiedAddress &uaddr) const
 {
-    // TODO
-    return AddressNotFound;
+    auto hdChain = m_wallet->GetMnemonicHDChain();
+    assert(hdChain.has_value());
+
+    auto ufvkid = m_wallet->FindUnifiedFullViewingKey(uaddr);
+    if (ufvkid.has_value()) {
+        // Look through the UFVKs that we have generated, and confirm that the
+        // seed fingerprint for the key we find for the ufvkid corresponds to
+        // the wallet's mnemonic seed.
+        for (const auto& [k, v] : m_wallet->mapUnifiedAccountKeys) {
+            if (v == ufvkid.value() && k.first == hdChain.value().GetSeedFingerprint()) {
+                return PaymentAddressSource::MnemonicHDSeed;
```

Comment:
We don't currently expose these payment sources in `listaddresses`. Do you want to do that as part of this PR or a separate PR?

Reply:
I think that should be a separate PR; I opened the issue for that yesterday as https://github.com/zcash/zcash/issues/5515